acl/policy.go

```diff
@@ -819,7 +821,17 @@ func TranslateLegacyRules(policyBytes []byte) ([]byte, error) {
 	rewritten := ast.Walk(parsed, func(node ast.Node) (ast.Node, bool) {
```

Comment:
It occurred to me that this walk is too general. If someone is unfortunate to have a segment value that happens to match one of these top level resource names like `key` or `session` it will mangle the translated rules:
```
consul (bugfix/5493)$ git diff
diff --git a/acl/policy_test.go b/acl/policy_test.go
index 8f5a100a8..e9191acb6 100644
--- a/acl/policy_test.go
+++ b/acl/policy_test.go
@@ -1422,6 +1422,9 @@ func TestRulesTranslate_GH5493(t *testing.T) {
                },
                "privatething2/": {
                        "policy": "deny"
+               },
+               "key": {
+                       "policy": "deny"
                }
        },
        "session": {
@@ -1472,6 +1475,10 @@ key_prefix "privatething2/" {
   policy = "deny"
 }
 
+key_prefix "key/" {
+  policy = "deny"
+}
+
 session_prefix "" {
   policy = "write"
 }

consul (bugfix/5493)$ go test ./acl -run TestRulesTranslate
--- FAIL: TestRulesTranslate_GH5493 (0.00s)
    require.go:157: 
        	Error Trace:	policy_test.go:1508
        	Error:      	Not equal: 
        	            	expected: "key_prefix \"\" {\n  policy = \"read\"\n}\n\nkey_prefix \"privatething1/\" {\n  policy = \"deny\"\n}\n\nkey_prefix \"anapplication/private/\" {\n  policy = \"deny\"\n}\n\nkey_prefix \"privatething2/\" {\n  policy = \"deny\"\n}\n\nkey_prefix \"key/\" {\n  policy = \"deny\"\n}\n\nsession_prefix \"\" {\n  policy = \"write\"\n}\n\nnode_prefix \"\" {\n  policy = \"read\"\n}\n\nagent_prefix \"\" {\n  policy = \"read\"\n}\n\nservice_prefix \"\" {\n  policy = \"read\"\n}\n\nevent_prefix \"\" {\n  policy = \"read\"\n}\n\nquery_prefix \"\" {\n  policy = \"read\"\n}"
        	            	actual  : "key_prefix \"\" {\n  policy = \"read\"\n}\n\nkey_prefix \"privatething1/\" {\n  policy = \"deny\"\n}\n\nkey_prefix \"anapplication/private/\" {\n  policy = \"deny\"\n}\n\nkey_prefix \"privatething2/\" {\n  policy = \"deny\"\n}\n\nkey_prefix key_prefix {\n  policy = \"deny\"\n}\n\nsession_prefix \"\" {\n  policy = \"write\"\n}\n\nnode_prefix \"\" {\n  policy = \"read\"\n}\n\nagent_prefix \"\" {\n  policy = \"read\"\n}\n\nservice_prefix \"\" {\n  policy = \"read\"\n}\n\nevent_prefix \"\" {\n  policy = \"read\"\n}\n\nquery_prefix \"\" {\n  policy = \"read\"\n}"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -16,3 +16,3 @@
        	            	 
        	            	-key_prefix "key/" {
        	            	+key_prefix key_prefix {
        	            	   policy = "deny"
        	Test:       	TestRulesTranslate_GH5493
FAIL
FAIL	github.com/hashicorp/consul/acl	0.002s
```

Reply:
I think this should now be taken care of.

An ObjectItem can have a list of keys so

```
key "key" {
   policy = "read"
```

and

```
{
   "key" : {
      "key": {
         "policy": "read"
      }
   }
}
```
end up using a single ObjectItem with a key length of 2. Whenever we encounter one of these we only ever modify the first key which should prevent modifying the resource segment.